drake/systems/plants/shapes/HeightMapTerrain.cpp

```diff
@@ -0,0 +1,212 @@
+#include "HeightMapTerrain.h"
+#include <iostream>
+#include <fstream>
+#include <cstring>
+
+using namespace std;
+using namespace Eigen;
+
+namespace DrakeShapes {
+// TODO: it would probably cleaner to have a struct per data type, so
+//  you could lookup byte sizes, conversion functions, etc.
+
+bool is_power_of_2(int x) { return (x > 0 && ((x & (x - 1)) == 0)); }
+
+HeightMapTerrain::HeightMapTerrain(const std::string &name,
+                                   const Eigen::Vector2i &ncells,
+                                   const Eigen::Vector2d &size)
+    : Geometry(HEIGHT_MAP_TERRAIN),
+      name(name),
+      ncells(ncells),
+      size(size),
+      m_gridHeightScale(1.0),
+      m_upAxis(2),
+      m_type(FLOAT) {
+  // check if ncells is of the form 2^N
+  assert(is_power_of_2(ncells(0)) && "ncells must be a power of 2.");
+  assert(is_power_of_2(ncells(1)) && "ncells must be a power of 2.");
+
+  nnodes = ncells.array() + 1;
+  nTotNodes = ((long)nnodes(0)) * nnodes(1);
+  bytesPerElement = getByteSize(m_type);
+  nBytes = nTotNodes * bytesPerElement;
+
+  m_rawHeightfieldData = new byte_t[nBytes];
+  if (m_rawHeightfieldData == NULL) assert(!"Out of memory");
+
+  delta_ell = size.array() / ncells.array().cast<double>();
+
+  fname = name + ".obj";
+}
+
+HeightMapTerrain::HeightMapTerrain(const HeightMapTerrain &other)
+    : Geometry(HEIGHT_MAP_TERRAIN) {
+  nBytes = other.nBytes;
+  nTotNodes = other.nTotNodes;
+  bytesPerElement = other.bytesPerElement;
+  size = other.size;
+  delta_ell = other.delta_ell;
+  ncells = other.ncells;
+  nnodes = other.nnodes;
+  m_gridHeightScale = other.m_gridHeightScale;
+  m_minHeight = other.m_minHeight;
+  m_maxHeight = other.m_maxHeight;
+  m_upAxis = other.m_upAxis;
+  m_type = other.m_type;
+  name = other.name;
+  fname = other.fname;
+
+  m_rawHeightfieldData = new byte_t[nBytes];
+  if (m_rawHeightfieldData == NULL) assert(!"Out of memory");
```

Comment:
This can't work because `new` will throw a `std::bad_alloc` exception; the next line won't get executed. In any case `assert()` is wrong here because (a) it should be used only to check programming invariants, and (b) it is only present in Debug builds. Running out of memory is a runtime problem that has to be dealt with.